### ISSUE TITLE
KeyError: u'wrapper_class'

### ISSUE THREAD

**Issue by samkuehn** [maraujop/django-crispy-forms#511]
Since upgrade to 1.5.1 I have been getting a KeyError: u'wrapper_class' under two circumstance.

1)  If I delete a form field without also deleting the corresponding layout field.  For instance:
```python
class MyForm(forms.From):
    my_field = forms.CharField()

    helper = FormHelper()
    helper.form_tag = False
    helper.layout = Layout(
        Field('my_field'),
    )

    def __init__(self, *args, **kwargs):
        if some_condition:
            self.fields.pop('my_field')
            # Must add the following line or you will get KeyError: u'wrapper_class'
            self.helper.layout.pop(0)
```
It can be a pain to find the exact field to pop in a more complex form (sometimes will look like this self.helper.layout[1][2][1].pop(2).  When changing the form layout it is easy to forget that you need to update the indexes of the layout field that you are deleting.  This is not a deal breaker but is a regression form 1.4.x.   At least there is a way to work around it :)

2)  The second circumstance is more sinister and is proving extremely difficult to pin down.   If you have a multiple crispy forms on the same page:
```
<form>
    {% crispy form1 %}
    {% crispy form2 %}
</form>
```
You will occasionally and seemingly randomly get KeyError: u'wrapper_class' in production (have not seen it while running the local dev server).  These are regular forms created in the view's context - no fields have been manually deleted as in the above example.  Restarting the gunicorn and memcached seems to temporarily "fix" the problem.  I know this will be difficult to track down given my inability to consistently reproduce the issue but I was hoping somebody else may have come across this.

Stack trace:
```
KeyError: u'wrapper_class'
  File "django/core/handlers/base.py", line 164, in get_response
    response = response.render()
  File "django/template/response.py", line 158, in render
    self.content = self.rendered_content
  File "django/template/response.py", line 135, in rendered_content
    content = template.render(context, self._request)
  File "django/template/backends/django.py", line 74, in render
    return self.template.render(context)
  File "django/template/base.py", line 209, in render
    return self._render(context)
  File "newrelic/api/function_trace.py", line 98, in dynamic_wrapper
    return wrapped(*args, **kwargs)
  File "django/template/base.py", line 201, in _render
    return self.nodelist.render(context)
  File "django/template/base.py", line 903, in render
    bit = self.render_node(node, context)
  File "django/template/base.py", line 917, in render_node
    return node.render(context)
  File "django/template/loader_tags.py", line 135, in render
    return compiled_parent._render(context)
  File "newrelic/api/function_trace.py", line 98, in dynamic_wrapper
    return wrapped(*args, **kwargs)
  File "django/template/base.py", line 201, in _render
    return self.nodelist.render(context)
  File "django/template/base.py", line 903, in render
    bit = self.render_node(node, context)
  File "django/template/base.py", line 917, in render_node
    return node.render(context)
  File "django/template/loader_tags.py", line 135, in render
    return compiled_parent._render(context)
  File "newrelic/api/function_trace.py", line 98, in dynamic_wrapper
    return wrapped(*args, **kwargs)
  File "django/template/base.py", line 201, in _render
    return self.nodelist.render(context)
  File "django/template/base.py", line 903, in render
    bit = self.render_node(node, context)
  File "django/template/base.py", line 917, in render_node
    return node.render(context)
  File "django/template/loader_tags.py", line 135, in render
    return compiled_parent._render(context)
  File "newrelic/api/function_trace.py", line 98, in dynamic_wrapper
    return wrapped(*args, **kwargs)
  File "django/template/base.py", line 201, in _render
    return self.nodelist.render(context)
  File "django/template/base.py", line 903, in render
    bit = self.render_node(node, context)
  File "django/template/base.py", line 917, in render_node
    return node.render(context)
  File "django/template/loader_tags.py", line 135, in render
    return compiled_parent._render(context)
  File "django/template/base.py", line 917, in render_node
    return node.render(context)
  File "django/template/defaulttags.py", line 329, in render
    return nodelist.render(context)
.. frames 26 until 41 were omitted and not available ..
  File "django/template/base.py", line 903, in render
    bit = self.render_node(node, context)
  File "django/template/base.py", line 917, in render_node
    return node.render(context)
  File "crispy_forms/templatetags/crispy_forms_tags.py", line 212, in render
    c = self.get_render(context)
  File "crispy_forms/templatetags/crispy_forms_tags.py", line 132, in get_render
    actual_form.form_html = helper.render_layout(actual_form, node_context, template_pack=self.template_pack)
  File "crispy_forms/helper.py", line 293, in render_layout
    template_pack=template_pack
  File "crispy_forms/layout.py", line 138, in render
    return self.get_rendered_fields(form, form_style, context, template_pack, **kwargs)
  File "crispy_forms/layout.py", line 102, in get_rendered_fields
    for field in self.fields
  File "crispy_forms/layout.py", line 102, in <genexpr>
    for field in self.fields
  File "crispy_forms/utils.py", line 72, in render_field
    form, form_style, context, template_pack=template_pack,
  File "crispy_forms/layout.py", line 357, in render
    fields = self.get_rendered_fields(form, form_style, context, template_pack, **kwargs)
  File "crispy_forms/layout.py", line 102, in get_rendered_fields
    for field in self.fields
  File "crispy_forms/layout.py", line 102, in <genexpr>
    for field in self.fields
  File "crispy_forms/utils.py", line 72, in render_field
    form, form_style, context, template_pack=template_pack,
  File "crispy_forms/layout.py", line 357, in render
    fields = self.get_rendered_fields(form, form_style, context, template_pack, **kwargs)
  File "crispy_forms/layout.py", line 102, in get_rendered_fields
    for field in self.fields
  File "crispy_forms/layout.py", line 102, in <genexpr>
    for field in self.fields
  File "crispy_forms/utils.py", line 72, in render_field
    form, form_style, context, template_pack=template_pack,
  File "crispy_forms/layout.py", line 442, in render
    **kwargs
  File "crispy_forms/layout.py", line 102, in get_rendered_fields
    for field in self.fields
  File "crispy_forms/layout.py", line 102, in <genexpr>
    for field in self.fields
  File "crispy_forms/utils.py", line 159, in render_field
    return html
  File "crispy_forms/base.py", line 22, in __exit__
    del self.context[key]
  File "django/template/context.py", line 75, in __delitem__
    del self.dicts[-1][key]
```
![screen shot 2015-09-09 at 10 09 12 am](https://cloud.githubusercontent.com/assets/263189/9767358/11e3dc42-56db-11e5-9c25-ceaaef14839d.png)
![screen shot 2015-09-09 at 10 08 40 am](https://cloud.githubusercontent.com/assets/263189/9767365/1863101a-56db-11e5-8ec8-93f607d920e3.png)

As a workaround perhaps we could check if the key exists before deleting it?

**Comment by carltongibson** [comment]
HI @samkuehn — Thanks for the detailed report. Sorry for your pain!

Some questions: 

1. Could you try using `git bisect` to find the commit where this snuck in?
2. Is it only ever the `wrapper_class` Key that's missing?

> You will occasionally and seemingly randomly...

Argh! :-)

**Comment by samkuehn** [comment]
1. It looks like the bug was introduced here: cad080ec41f870734e3e01c58f7a13d156db7b1a. I was only able to test it under scenario 1 above (deleting form field w/out also deleting layout field).  Hopefully fixing this will fix scenario 2 as well.
2. Yes it is only ever the ```wrapper_class```

**Comment by carltongibson** [comment]
Good work thanks. 

> As a workaround perhaps we could check if the key exists before deleting it?

Looks like that's what used to happen. I'm curious about the Python 3.3 compatibility remark there. 

OK. For mega bonus points (and a merge + release first thing) do you fancy opening a PR resolving this? (I would say just revert — but I'm guessing _something_ will fail when you do that...)

**Comment by carltongibson** [comment]
Fixed in 1.5.2 (available on PyPI now)
